Subject: [PATCH] Store package restriction settings at shutdown

Symptom:
Disabled package was re-enabled by restarting a device.

Root cause:
PMS postpones storing the package restriction settings in 10
seconds by design. When a package gets disabled right before
shutdown, PMS fails to store the last changes.

Solution:
Store package restriction settings at shutdown.

diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index 4454c71..9826027 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -10294,6 +10294,12 @@
         mPackageUsage.writeNow(mPackages);
         mCompilerStats.writeNow();
         mDexManager.writePackageDexUsageNow();
+
+        // This is the last chance to write out pending restriction settings.
+        if (mHandler.hasMessages(WRITE_PACKAGE_RESTRICTIONS)) {
+            mHandler.removeMessages(WRITE_PACKAGE_RESTRICTIONS);
+            mHandler.sendEmptyMessage(WRITE_PACKAGE_RESTRICTIONS);
+        }
     }
 
     @Override
